server/src/main/java/org/elasticsearch/index/shard/UnpromotableRefreshService.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.shard;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionListenerResponseHandler;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.admin.indices.refresh.TransportUnpromotableShardRefreshAction;
+import org.elasticsearch.action.admin.indices.refresh.UnpromotableShardRefreshRequest;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+public class UnpromotableRefreshService {
```

Comment:
I am not sure why this needs to be in core and also not why it needs to be passed in through the engine config when it only ends up in the archiver?

Can it be moved to the stateless module instead?

Reply:
Yes. I was wondering the same myself. Please see the first comments in the PR, https://github.com/elastic/elasticsearch/pull/95331#issuecomment-1513297864, https://github.com/elastic/elasticsearch/pull/95331#issuecomment-1514508409. Didn't find an obvious way of doing this. But I'm all for changing this. Had some ideas, but didn't want to spend too much time on the wiring part of the PR(s).